# Lab 7: Investing time in functions and data structures

Welcome to the seventh lab of the course; we will do exercises on financial investments exploiting functions and data structures.

**DISCLAIMERS**
- No investing recommendations are given here
- Implemented procedures are exemplified and not guaranteed to be accurate

## Exercise 1: Interests

Create a function called 'calculate_interests' that takes in input:
- 'capital': the amount of money you invest
- 'interest_as_percentage': a number that indicates the percentage of interests
and returns the amount of money earned.

For instance,

    calculate_interests(1000,2)

returns 20.0 (i.e., the 2% of 1000).


In [9]:
def calculate_interests(capital,interest_as_percentage):
    # TODO EXERCISE

In [8]:
calculate_interests(1000,2)

20.0

## Exercise 2: Compound interests

Interests must be applied every year. Define a function 'invest' that takes in input:
- 'capital': the amount of money you invest
- 'yearly_interest_as_percentage': the interest to be applied **every year**
- 'number_of_year': how many years you plan to invest
and *prints*, year after year, the amount of money earner and the new capital; use 'round(num,2)' to round a number 'num' to 2 decimals.

For instance,

    invest(1000,1,5)

returns

> You start with 1000  
> In the year 1 you have earned 10.0; you now have 1010.0  
> In the year 2 you have earned 10.1; you now have 1020.1  
> In the year 3 you have earned 10.2; you now have 1030.3  
> In the year 4 you have earned 10.3; you now have 1040.6  
> In the year 5 you have earned 10.41; you now have 1051.

What does it take to get this output?

In [39]:
def invest(capital, yearly_interest_as_percentage, number_of_years):
    # TODO EXERCISE

In [40]:
invest(1000,1,5)

You start with 1000
In the year 1 you have earned 10.0; you now have 1010.0
In the year 2 you have earned 10.1; you now have 1020.1
In the year 3 you have earned 10.2; you now have 1030.3
In the year 4 you have earned 10.3; you now have 1040.6
In the year 5 you have earned 10.41; you now have 1051.01


## Exercise 3: Calculate the profit

Define a function 'invest2' that works as above, but also calculates the profit at the end of the investment.

For instance,

    invest2(1000,1,5)

returns the following (only the last line is different from Exercise 2)

> You start with 1000  
> In the year 1 you have earned 10.0; you now have 1010.0  
> In the year 2 you have earned 10.1; you now have 1020.1  
> In the year 3 you have earned 10.2; you now have 1030.3  
> In the year 4 you have earned 10.3; you now have 1040.6  
> In the year 5 you have earned 10.41; you now have 1051.01
> In the end, your profit is 51.01

What does it take to get this output?

In [41]:
def invest2(capital, yearly_interest_as_percentage, number_of_years):
    # TODO EXERCISE

In [42]:
invest2(1000,1,5)

You start with 1000
In the year 1 you have earned 10.0; you now have 1010.0
In the year 2 you have earned 10.1; you now have 1020.1
In the year 3 you have earned 10.2; you now have 1030.3
In the year 4 you have earned 10.3; you now have 1040.6
In the year 5 you have earned 10.41; you now have 1051.01
In the end, your profit is 51.01


## Exercise 4: Taxes

Profits must be taxed. Define a function 'remove_taxes' that takes in input:
- 'amount': an amount of money
- 'tax_percentage': the percentage of taxes to apply
and returns the amount without taxes.

For instance,

    remove_taxes(1000,26)

returns 740.0 (i.e., 1000-26%).

In [22]:
def remove_taxes(amount, tax_percentage):
    # TODO EXERCISE

In [58]:
remove_taxes(1000,26)

740.0

## Exercise 5: Include taxes in the investment

Define a function 'invest3' that works as 'invest2', but also removes taxes from the profit. The percentage of taxes must be given in input, with a default value of 26. **Notice**: taxes are applied only if the profit is positive! 

For instance,

    invest3(1000,1,5)

returns the following

> You start with 1000  
> In the year 1 you have earned 10.0; you now have 1010.0  
> In the year 2 you have earned 10.1; you now have 1020.1  
> In the year 3 you have earned 10.2; you now have 1030.3  
> In the year 4 you have earned 10.3; you now have 1040.6  
> In the year 5 you have earned 10.41; you now have 1051.01  
> Your gross profit is 51.01  
> After taxes, your net profit is 37.75, so you end up with a total of 1037.75

In [45]:
def invest3(capital, yearly_interest_as_percentage, number_of_years, tax_percentage=26):
    # TODO EXERCISE

In [46]:
invest3(1000,1,5)

You start with 1000
In the year 1 you have earned 10.0; you now have 1010.0
In the year 2 you have earned 10.1; you now have 1020.1
In the year 3 you have earned 10.2; you now have 1030.3
In the year 4 you have earned 10.3; you now have 1040.6
In the year 5 you have earned 10.41; you now have 1051.01
Your gross profit is 51.01
After taxes, your net profit is 37.75, so you end up with a total of 1037.75


## Exercise 6: Include costs in the investment

Usually, investments have costs, that must be considered every year. The percentage of costs must be given in input, with a default value of 2. **Notice**: every year, costs are calculated based on the value of capital+interests in such year.

Define a function 'invest4' that works like 'invest3' but also includes the costs. For instance,

    invest4(1000,1,5)

returns the following

> You start with 1000  
> In the year 1 you have earned 10.0 and paid 5.05; you now have 1004.95  
> In the year 2 you have earned 10.05 and paid 5.07; you now have 1009.92  
> In the year 3 you have earned 10.1 and paid 5.1; you now have 1014.92  
> In the year 4 you have earned 10.15 and paid 5.13; you now have 1019.95  
> In the year 5 you have earned 10.2 and paid 5.15; you now have 1025.0  
> Your gross profit is 25.0  
> After taxes, your net profit is 18.5, so you end up with a total of 1018.5  

In [59]:
def invest4(capital, yearly_interest_as_percentage, number_of_years, tax_percentage=26, yearly_cost_as_percentage=2):
    # TODO EXERCISE

In [60]:
invest4(1000,1,5,yearly_cost_as_percentage=0.5)

You start with 1000
In the year 1 you have earned 10.0 and paid 5.05; you now have 1004.95
In the year 2 you have earned 10.05 and paid 5.07; you now have 1009.92
In the year 3 you have earned 10.1 and paid 5.1; you now have 1014.92
In the year 4 you have earned 10.15 and paid 5.13; you now have 1019.95
In the year 5 you have earned 10.2 and paid 5.15; you now have 1025.0
Your gross profit is 25.0
After taxes, your net profit is 18.5, so you end up with a total of 1018.5


## Exercise 7: Variable interests

Interests change every year, and they are not always positive. Define a function 'invest5' that works like 'invest4', but where:
- the parameter 'yearly_interest_as_percentage' is replaced with 'yearly_interests_as_percentages' as a list of interests
- the parameter 'number_of_years' is not needed anymore (it is implict in the length of 'yearly_interests_as_percentages')

For instance,

    invest5(1000,[1,-2,3,-4,5],yearly_cost_as_percentage=0.5)

returns the following

> You start with 1000
> In the year 1 you have earned 10.0 and paid 5.05; you now have 1004.95  
> In the year 2 you have earned -20.1 and paid 4.92; you now have 979.93  
> In the year 3 you have earned 29.4 and paid 5.05; you now have 1004.28  
> In the year 4 you have earned -40.17 and paid 4.82; you now have 959.29  
> In the year 5 you have earned 47.96 and paid 5.04; you now have 1002.21  
> Your gross profit is 2.21  
> After taxes, your net profit is 1.64, so you end up with a total of 1001.64

In [64]:
def invest5(capital, yearly_interests_as_percentages, tax_percentage=26, yearly_cost_as_percentage=2):
    # TODO EXERCISE

In [65]:
invest5(1000,[1,-2,3,-4,5],yearly_cost_as_percentage=0.5)

You start with 1000
In the year 1 you have earned 10.0 and paid 5.05; you now have 1004.95
In the year 2 you have earned -20.1 and paid 4.92; you now have 979.93
In the year 3 you have earned 29.4 and paid 5.05; you now have 1004.28
In the year 4 you have earned -40.17 and paid 4.82; you now have 959.29
In the year 5 you have earned 47.96 and paid 5.04; you now have 1002.21
Your gross profit is 2.21
After taxes, your net profit is 1.64, so you end up with a total of 1001.64


## Exercise 8: Make it a dict

Instead of just printing data, put everything into a dictionary structured as follows.

    {
        initialInvestment: 1000
        yearlyInvestments: [
            {initialCapital: 1000, year: 1, interest_perc: 1, interests: 10.0, cost: 5.05, newCapital: 1004.95}
            ...    
        ]
        grossProfit: 2.21
        netProfit: 1.64
        finalAmount: 1001.64
    }

In [ ]:
def invest6(capital, yearly_interests_as_percentages, tax_percentage=26, yearly_cost_as_percentage=2):
    # TODO EXERCISE